src/lib/flow/sol-flow-js.c

```diff
@@ -0,0 +1,1422 @@
+}
+
+static duk_ret_t
+send_string_packet(struct sol_flow_node *node, uint16_t port, duk_context *ctx)
+{
+    const char *value = duk_require_string(ctx, 1);
+
+    sol_flow_send_string_packet(node, port, value);
+
+    return 0;
+}
+
+static struct sol_flow_node *
+get_node_from_duk_ctx(duk_context *ctx)
+{
+    struct sol_flow_node *n;
+
+    duk_push_global_object(ctx);
+
+    duk_get_prop_string(ctx, -1, "\xFF" "Soletta_node_pointer");
+    n = duk_require_pointer(ctx, -1);
+
+    duk_pop_2(ctx); /* Soletta_node_pointer, global object values */
+
+    return n;
+}
+
+static int
+get_output_port_number(const struct flow_js_type *type, const char *port_name)
+{
+    struct flow_js_port_out *p;
+    uint16_t i;
+
+    SOL_VECTOR_FOREACH_IDX (&type->ports_out, p, i) {
+        if (streq(p->name, port_name))
+            return i;
+    }
+
+    return -EINVAL;
+}
+
+static duk_ret_t
+send_packet(duk_context *ctx)
+{
+    const struct flow_js_port_out *port;
+    const struct flow_js_type *type;
+    const char *port_name;
+    struct sol_flow_node *node;
+    int port_number;
+
+    port_name = duk_require_string(ctx, 0);
+
+    node = get_node_from_duk_ctx(ctx);
+    type = (struct flow_js_type *)node->type;
+    if (!node || !type) {
+        SOL_WRN("Couldn't send packet to '%s' port.", port_name);
+        return 0;
+    }
+
+    /* TODO: Check a cheaper way to do this, if we had hashes we could
+     * use them here. */
+
+    port_number = get_output_port_number(type, port_name);
+    if (port_number < 0) {
+        SOL_ERR("'%s' invalid port name.", port_name);
+        return 0;
+    }
+
+    port = sol_vector_get(&type->ports_out, port_number);
+    if (!port) {
+        SOL_ERR("'%s' invalid port name.", port_name);
+        return 0;
+    }
+
+    if (streq(port->type_name, "boolean"))
+        return send_boolean_packet(node, port_number, ctx);
+    if (streq(port->type_name, "byte"))
+        return send_byte_packet(node, port_number, ctx);
+    if (streq(port->type_name, "float"))
+        return send_float_packet(node, port_number, ctx);
+    if (streq(port->type_name, "int"))
+        return send_int_packet(node, port_number, ctx);
+    if (streq(port->type_name, "rgb"))
+        return send_rgb_packet(node, port_number, ctx);
+    if (streq(port->type_name, "string"))
+        return send_string_packet(node, port_number, ctx);
+
+    /* TODO: Create a way to let the user define custom packets. Maybe we could
+     * use the same techniques we do for option parsing, and provide an object
+     * with an array of fields, offsets and values in basic C types. */
+
+    SOL_WRN("Couldn't handle unknown port type %s.", port->type_name);
+
+    return 0;
+}
+
+static duk_ret_t
+send_error_packet(duk_context *ctx)
+{
+    const char *value_msg = NULL;
+    struct sol_flow_node *node;
+    int value_code;
+
+    value_code = duk_require_int(ctx, 0);
+
+    if (duk_is_string(ctx, 1))
+        value_msg = duk_require_string(ctx, 1);
```

Comment:
Instead of checking if the variable is a string, can't Duktape just coerce the value into a string automatically? (In most cases it's going to be useless, though; objects in JavaScript are often coerced to "[object Object]", which isn't that great for debugging purposes, but a second number might be.)